src/Models/NonhydrostaticModels/immersed_pressure_solver.jl

```diff
@@ -0,0 +1,152 @@
+                               abstol = 0,
+                               kw...)
+
+    if preconditioner == "FFT"
+        arch = architecture(grid)
+        preconditioner = PressureSolver(arch, grid)
+    end
+
+    rhs = CenterField(grid)
+
+    pcg_solver = PreconditionedConjugateGradientSolver(compute_laplacian!; reltol, abstol,
+                                                       preconditioner,
+                                                       template_field = rhs,
+                                                       kw...)
+
+    return ImmersedPoissonSolver(rhs, grid, pcg_solver)
+end
+
+@kernel function calculate_pressure_source_term!(rhs, grid, Δt, U★)
+    i, j, k = @index(Global, NTuple)
+    @inbounds rhs[i, j, k] = divᶜᶜᶜ(i, j, k, grid, U★.u, U★.v, U★.w) / Δt
+end
+
+@inline laplacianᶜᶜᶜ(i, j, k, grid, ϕ) = ∇²ᶜᶜᶜ(i, j, k, grid, ϕ)
+
+@kernel function laplacian!(∇²ϕ, grid, ϕ)
+    i, j, k = @index(Global, NTuple)
+    @inbounds ∇²ϕ[i, j, k] = laplacianᶜᶜᶜ(i, j, k, grid, ϕ)
+end
+
+function compute_laplacian!(∇²ϕ, ϕ)
+    grid = ϕ.grid
+    arch = architecture(grid)
+
+    fill_halo_regions!(ϕ)
+
+    launch!(arch, grid, :xyz, laplacian!, ∇²ϕ, grid, ϕ)
+
+    return nothing
+end
+
+function solve_for_pressure!(pressure, solver::ImmersedPoissonSolver, Δt, U★)
+    # TODO: Is this the right criteria?
+    min_Δt = eps(typeof(Δt))
+    Δt <= min_Δt && return pressure
+
+    rhs = solver.rhs
+    grid = solver.grid
+    arch = architecture(grid)
+
+    if grid isa ImmersedBoundaryGrid
+        underlying_grid = grid.underlying_grid
+    else
+        underlying_grid = grid
+    end
+
+    launch!(arch, grid, :xyz, calculate_pressure_source_term!,
+            rhs, underlying_grid, Δt, U★)
+
+    mask_immersed_field!(rhs, zero(grid))
+
+    # Solve pressure Pressure equation for pressure, given rhs
+    # @info "Δt before pressure solve: $(Δt)"
+    solve!(pressure, solver.pcg_solver, rhs)
+
+    return pressure
+end
+
+struct MITgcmPreconditioner end
+
+@inline function precondition!(P_r, ::MITgcmPreconditioner, r, args...)
+    grid = r.grid
+    arch = architecture(P_r)
+
+    fill_halo_regions!(r)
+
+    launch!(arch, grid, :xyz, _MITgcm_precondition!,
+            P_r, grid, r)
+
+    return P_r
+end
+
+# Kernels that calculate coefficients for the preconditioner
+@inline Ax⁻(i, j, k, grid) = @inbounds Axᶠᶜᶜ(i, j, k, grid) / Δxᶠᶜᶜ(i, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
+@inline Ay⁻(i, j, k, grid) = @inbounds Ayᶠᶜᶜ(i, j, k, grid) / Δyᶠᶜᶜ(i, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
+@inline Az⁻(i, j, k, grid) = @inbounds Azᶠᶜᶜ(i, j, k, grid) / Δzᶠᶜᶜ(i, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
+@inline Ax⁺(i, j, k, grid) = @inbounds Axᶠᶜᶜ(i+1, j, k, grid) / Δxᶠᶜᶜ(i+1, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
+@inline Ay⁺(i, j, k, grid) = @inbounds Ayᶠᶜᶜ(i, j+1, k, grid) / Δyᶠᶜᶜ(i, j+1, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
+@inline Az⁺(i, j, k, grid) = @inbounds Azᶠᶜᶜ(i, j, k+1, grid) / Δzᶠᶜᶜ(i, j, k+1, grid) / Vᶜᶜᶜ(i, j, k, grid)
```

Comment:
```suggestion
@inline Ax⁻(i, j, k, grid) = Axᶠᶜᶜ(i, j, k, grid) / Δxᶠᶜᶜ(i, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
@inline Ay⁻(i, j, k, grid) = Ayᶠᶜᶜ(i, j, k, grid) / Δyᶠᶜᶜ(i, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
@inline Az⁻(i, j, k, grid) = Azᶠᶜᶜ(i, j, k, grid) / Δzᶠᶜᶜ(i, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
@inline Ax⁺(i, j, k, grid) = Axᶠᶜᶜ(i+1, j, k, grid) / Δxᶠᶜᶜ(i+1, j, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
@inline Ay⁺(i, j, k, grid) = Ayᶠᶜᶜ(i, j+1, k, grid) / Δyᶠᶜᶜ(i, j+1, k, grid) / Vᶜᶜᶜ(i, j, k, grid)
@inline Az⁺(i, j, k, grid) = Azᶠᶜᶜ(i, j, k+1, grid) / Δzᶠᶜᶜ(i, j, k+1, grid) / Vᶜᶜᶜ(i, j, k, grid)
```

Reply:
btw, why all these are computed at `fcc`, what's so special about `fcc`? I don't see that..

Reply:
Good catch! Let me change it and test it, will update soon.